src/template/js/sdk/Capabilities/index.mjs

```diff
@@ -18,43 +18,82 @@
+  }
 
-const getProviderArgs = function(...args) {
-  const provider = args.pop()
-  const module = args[0].toLowerCase() || '*'
-  return [module, provider]
+  methods.forEach(m => m.name = `${module}.${m.name}`)
+  providerInterfaces[capability] = methods.concat()
 }
 
-const provide = function(...args) {
-  const [module, provider] = getProviderArgs(...args)
-  const methods = Object.getOwnPropertyNames(Object.getPrototypeOf(provider)).filter(item => {
-    return typeof provider[item] === 'function' && item != 'constructor'
-  })
-  const pms = validProviderMethods[module]
-  for (let i = 0; i < methods.length; i++) {
-    const name = methods[i].charAt(0).toUpperCase() + methods[i].slice(1);
-    if (pms.indexOf(methods[i]) !== -1) {
-      Events.listen(module, 'request' + name, async function (req) {
-        const fn = provider[methods[i]]
-        const providerCallArgs = [req.request, resp => {
-          Transport.send(module, methods[i] + 'Response', {
-            correlationId: req.correlationId,
-            response: resp
+const provide = function(capability, provider) {
+  const methods = []
+  const iface = providerInterfaces[capability]
+
+  if (provider.constructor.name !== 'Object') {
+    methods.push(...Object.getOwnPropertyNames(Object.getPrototypeOf(provider)).filter(item => typeof provider[item] === 'function' && item !== 'constructor'))
+  }
+  else {
+    methods.push(...Object.getOwnPropertyNames(provider).filter(item => typeof provider[item] === 'function'))
+  }
+
+  if (iface) {
+
+    // make sure every interfaced method exists in the providers methods list
+    const valid = iface.every(method => methods.find(m => m === method.name.split('.').pop()))
+
+    if (valid) {
+      iface.forEach(imethod => {
+        const parts = imethod.name.split('.')
+        const method = parts.pop();
+        const module = parts.pop().toLowerCase();
+
+        const defined = !!methods.find(m => m === method) 
+
+        if (defined) {
+          Events.listen(module, `request${method.charAt(0).toUpperCase() + method.substr(1)}`, function (request) {
+
+            const providerCallArgs = []
+            
+            // only pass in parameters object if schema exists
+            if (imethod.parameters) {
+              providerCallArgs.push(request.parameters)
+            }
+
+            // only pass in the ready handshake if needed
+            if (imethod.handshake) {
+              providerCallArgs.push( _ => {
+                Transport.send(module, `${method}Ready`, {
+                  correlationId: request.correlationId
+                })
+              })
+            }
+
+            const response = {
+              correlationId: request.correlationId
+            }
+
+            const result = provider[method].apply(provider, providerCallArgs).then(result => {
+              if (imethod.response) {
+                response.result = result
```

Comment:
did you change the name of response to result?

Reply:
Yes, to be consistent with the `polymorphic-pull` pattern.

We used request/response for the RPC objects that contain correlationId, etc.
We used parameters/result (borrowed from OpenRPC's conventions) for the actual things passed back and forth to the SDK.

Reply:
maybe it is fine, seems weird to use result when that is a term in JSONRPC. Quick glances, make it seem like a JSONRPC response instead of a request.

Reply:
hmm, interesting point. Maybe we revisit when we go to 1.0? Too late to change `polymorphic-pull`